lib/elixir/lib/module.ex

```diff
@@ -1097,14 +1138,197 @@ defmodule Module do
         :ok
       {:error, :private_doc} ->
         :elixir_errors.warn line, env.file,
-          "function #{name}/#{arity} is private, " <>
+          "function #{format_function(pair)} is private, " <>
           "@doc's are always discarded for private functions"
     end
 
     :ok
   end
 
   @doc false
+  # Used internally to check impl usage.
+  # This function is private and must be used only internally.
+  def check_impl(env, kind, name, args, _guards, _body) do
+    %{module: module, line: line, file: file} = env
+
+    table = data_table_for(module)
+    value = get_impl_info(table)
+
+    if value do
+      set_impl_required(table, line)
+      set_doc_false_if_unset(table, line)
+    end
+
+    arity = length(args)
+    pair = {name, arity}
+    required? = impl_required_for_callbacks?(table)
+
+    case do_check_impl(module, kind, pair, value, required?) do
+      {:error, message} -> :elixir_errors.warn line, file, message
+      :ok -> :ok
+    end
+  end
+
+  defp set_impl_required(table, unread_line) do
+    case :ets.lookup(table, {:elixir, :impl_required}) do
+      [{{:elixir, :impl_required}, _value, _accumulated?, _unread_line}] -> :ok
+      _ ->
+        :ets.insert(table, {{:elixir, :impl_required}, true, false, unread_line})
+        :ok
+    end
+  end
+
+  defp set_doc_false_if_unset(table, line) do
+    case :ets.lookup(table, :doc) do
+      [{:doc, {_line, <<_::binary>>}, _accumulated?, _unread_line}] ->
+        :doc_exists
+      [{:doc, {_line, false}, _accumulated?, _unread_line}] ->
+        :doc_already_set_false
+      [] ->
+        :ets.insert(table, {:doc, {line, false}, false, line})
+    end
+  end
+
+  defp impl_required_for_callbacks?(table) do
+    case :ets.lookup(table, {:elixir, :impl_required}) do
+      [{{:elixir, :impl_required}, true, _accumulated?, _unread_line}] -> true
+      _ -> false
+    end
+  end
+
+  defp format_function({name, arity}), do: format_function(name, arity)
+  defp format_function(name, arity)
+       when (is_atom(name) or is_binary(name)) and is_integer(arity) do
+    "#{name}/#{arity}"
+  end
+
+  defp do_check_impl(_, kind, _, value, _)
+       when value == nil and kind in [:type, :typep, :defp, :defmacrop] do
+    :ok
+  end
+  defp do_check_impl(_, kind, function_tuple, _, _) when kind in [:defp, :defmacrop] do
+    {:error, "function #{format_function(function_tuple)} is private, @impl " <>
+             "is always ignored for private functions"}
+  end
+  defp do_check_impl(_, kind, _, _, _) when kind in [:type, :typep] do
+    {:error, "@impl is always ignored for type annotations"}
+  end
+  defp do_check_impl(_, _, _, value, impl_required_for_callbacks?)
+       when value in [nil, false] and not impl_required_for_callbacks? do
+    :ok
+  end
+  defp do_check_impl(module, _, function_tuple, value, impl_required_for_callbacks?)
+       when value in [nil, false] and impl_required_for_callbacks? do
```

Comment:
It might make sense to merge this somehow with the `true` and behaviour cases since they all duplicate this `callbacks_with_behaviour` logic.